extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java

```diff
@@ -0,0 +1,94 @@
+ * IdentityHub instance and verifies the credentials.
+ */
+public class IdentityHubCredentialsVerifier implements CredentialsVerifier {
+
+    private static final String IDENTITY_HUB_SERVICE_TYPE = "IdentityHub";
+    private final IdentityHubClient identityHubClient;
+    private final Monitor monitor;
+    private final JwtCredentialsVerifier jwtCredentialsVerifier;
+    private final VerifiableCredentialsJwtUtils jwtPayloadParser;
+
+    /**
+     * Create a new credential verifier that uses an Identity Hub
+     *
+     * @param identityHubClient IdentityHubClient.
+     */
+    public IdentityHubCredentialsVerifier(IdentityHubClient identityHubClient, Monitor monitor, JwtCredentialsVerifier jwtCredentialsVerifier, VerifiableCredentialsJwtUtils JWTPayloadParser) {
+        this.identityHubClient = identityHubClient;
+        this.monitor = monitor;
+        this.jwtCredentialsVerifier = jwtCredentialsVerifier;
+        this.jwtPayloadParser = JWTPayloadParser;
+    }
+
+    /**
+     * Get credentials from the IdentityHub of participant, verifies the credentials, and returns the successfully verified credentials.
+     * To be successfully verified, credentials needs to be signed by the specified issuer in the JWT.
+     *
+     * @param didDocument of a participant. The Did Document should contain an IdentityHub service.
+     * @return VerifiableCredentials.
+     */
+    @Override
+    public Result<Map<String, Object>> getVerifiedCredentials(DidDocument didDocument) {
+        var hubBaseUrl = getIdentityHubBaseUrl(didDocument);
+        if (hubBaseUrl.failed()) return Result.failure(hubBaseUrl.getFailureMessages());
+
+        var jwts = identityHubClient.getVerifiableCredentials(hubBaseUrl.getContent());
+        if (jwts.failed()) {
+            return Result.failure(jwts.getFailureMessages());
+        }
+        var verifiedClaims = jwts.getContent()
+                .stream()
+                .filter(jwt -> jwtCredentialsVerifier.verifyClaims(jwt, didDocument.getId()))
+                .filter(jwtCredentialsVerifier::isSignedByIssuer);
+
+        var credentials = verifiedClaims.map(jwtPayloadParser::extractCredential).collect(Collectors.toList());
+        credentials.stream().filter(AbstractResult::failed).forEach(result -> monitor.warning(String.join(",", result.getFailureMessages())));
+
+        var claims = credentials.stream().filter(AbstractResult::succeeded)
+                .map(AbstractResult::getContent)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        return Result.success(claims);
+    }
+
+    private Result<String> getIdentityHubBaseUrl(DidDocument didDocument) {
+        var hubBaseUrl = didDocument
+                .getService()
+                .stream()
+                .filter(s -> s.getType().equals(IDENTITY_HUB_SERVICE_TYPE))
+                .findFirst();
+
+        if (hubBaseUrl.isEmpty()) return Result.failure("Failed getting identityHub URL");
```

Comment:
these two lines can be replaced with
```
return hubBaseUrl
                .map(...)
                .orElse() // or .orElseGet()
```

Reply:
More elegant indeeed. I have adapted the test to verify the failure message as well.